### ISSUE TITLE
Add algorithm for overlay of geometries with a rectangular grid

### ISSUE THREAD

**Issue by dbaston** [libgeos/geos#967]
I am working on a project to expose the [exactextract](https://github.com/isciences/exactextract) library to Python in a way similar to the R package [exactextractr](https://github.com/isciences/exactextractr).

Instead of creating separate Python bindings for this library, I am considering attempting to move much of the code of this library into GDAL where it would become accessible through GDAL's Python bindings. Related but not necessarily dependent on this, I am thinking about whether some of the code could move into GEOS -- specifically, an optimized intersection calculation between rectangular grids and polygonal/linear geometries.

The result of the overlay is most commonly expressed as the fraction of each grid cell that is covered by a polygon, although it is also possible to get the total length of linear geometries within each grid cell or (more experimentally) a Geometry representation of each polygon/grid cell intersection, e.g.

![image](https://github.com/libgeos/geos/assets/6318931/e3e96bf3-8be0-409e-b8a9-9e7a5b3739a8)

This code has a strong test suite and has had pretty good public adoption since the R package was published in 2019, so I think it can be considered reasonably mature.

I don't have a detailed proposal yet, but I imagine that a minimal-footprint addition to GEOS would add something along these lines:

```
// define a grid
GEOSGrid* GEOSGrid_create(double xmin, double ymin, double xmax, double ymax, double dx, double dy);

// calculate a Grid coincident with `grid` but covering only `geom`
GEOSGrid* GEOSGrid_clip(const GEOSGrid* grid, const GEOSGeometry* geom);

// partition a polygonal geometry by a grid
GEOSGeometry* GEOSGrid_intersection(const GEOSGrid* grid, const GEOSGeometry* geom);

// calculate a raster with the fraction of each grid cell that is covered
// to avoid introducing a matrix type (and having slow access through non-inlined function calls)
// just return a float* and access values with x[i*cols + j] or something
float* GEOSGrid_intersectionAreas(const GEOSGrid*, const GEOSGeometry* geom);
```

Just thought I'd create an issue for this in case anyone has thoughts. I'm going to aim to have a better formed proposal before the Vienna code sprint.

**Comment by mwtoews** [comment]
The proposed GEOSGrid_create has xmin, xmax and dx as floats, which is potentially problematic since dx may not be a multiple of (xmax - ymin). How about something based on a grid shape with offset to the top left corner:
```
GEOSGrid* GEOSGrid_create(int nrow, int ncol, double dx, double dy, double ul_x, double ul_y);
```
These parameters can be evaluated from bounding box inputs or raster affine transformation coefficients.

**Comment by dr-jts** [comment]
W. R. Franklin proposed a clever algorithm for computing the area of intersection of two polygons without forming the actual geometry of intersection.  The advantage of this algorithm is it's simplicity and robustness.  There's a (mostly complete) implementation in a JTS branch: [`OverlayArea`](https://github.com/dr-jts/jts/blob/intersection-area/modules/lab/src/main/java/org/locationtech/jts/operation/overlayarea/OverlayArea.java)

Obviously it would be easy to adapt this to using a rectangle as one of the polygons.  Might this be worth looking at?

**Comment by dbaston** [comment]
> The proposed GEOSGrid_create has xmin, xmax and dx as floats, which is potentially problematic since dx may not be a multiple of (xmax - ymin).

Yes, there is a definite conflict between how people often talk about grids ("it goes from -180 to 180 at 10 arc-second resolution") and how the math works out. The approach I took is to accept the xmin/xmax based specification and allow the size of the rightmost column/bottom row to adjust to compensate. In the example above the right boundary of the rightmost cell would be adjusted from 179.99999999999994 to 180.

> How about something based on a grid shape with offset to the top left corner:

This would work, too.

> There is also the possibility of adding a rotation term too.

exactextract does not handle rotated rectangles, though I suppose the same approach could work. Just a trickier implementation.

> Obviously it would be easy to adapt this to using a rectangle as one of the polygons. Might this be worth looking at?

A primary advantage of exactextract is calculating the intersection area against all rectangles in a single pass, rather than performing m x n separate intersection calculations.

**Comment by dr-jts** [comment]
Would the GEOS API actually get used by 3rd parties?  Or is it just to support the `exactextract` lib whereever it ends up?

**Comment by dbaston** [comment]
I'd hope for it to be used by other GEOS users, otherwise there is no benefit to putting it into the library.

**Comment by dbaston** [comment]
Updated to support multipart polygons and holes. It takes about 5 seconds to subdivide the GADM level 0 boundaries (256 polygons, 32.8 million vertices) against 250,000 grid cell polygons:

![image](https://github.com/libgeos/geos/assets/6318931/96c29bfc-6db3-460d-9ded-cc855a47a40d)

I think the primary benefit in GEOS would be this polygon subdivision, but I would need an API to efficiently get the part I need (coverage percent) without materializing all of the result polygons.

**Comment by pramsey** [comment]
Very nice!

**Comment by dbaston** [comment]
Another possible interface -- return an "intersection result" that makes it possible to get either the overlap fractions or geometries either in bulk or individually by row/col.

```
GEOSGridIntersection* GEOSGrid_intersection(const GEOSGrid*, const GEOSGeometry* geom);
float* GEOSGridIntersection_getIntersectionAreas(const GEOSGridIntersection*);
float GEOSGridIntersection_getIntersectionArea(const GEOSGridIntersection*, size_t row, size_t col);
GEOSGeometry* GEOSGridIntersection_getIntersectionGeometries(const GEOSGridIntersection*);
GEOSGeometry* GEOSGridIntersection_getIntersectionGeometry(const GEOSGridIntersection*, size_t row, size_t col);
```

I can imagine using this to partition up work, so getting the intersection geometries in a structured way (rather than combined in a `GeometryCollection`) could be useful.

**Comment by pramsey** [comment]
Does row/col make sense? for a concave shape, there will be places where there is no geometry in some grid cells. Is the upside value of doing row/col worth the extra scruft of a new object to create/destroy?

**Comment by dbaston** [comment]
> there will be places where there is no geometry in some grid cells.

This is where row/column access could be useful, because otherwise there is no obvious correspondence between grid cells and result geometries. As an example, I was thinking about a point-in-polygon algorithm that first divides the space into, say, a 100x100 grid. To do to the PIP test, you calculate the cell associated with your point, check to see if that cell has a covered fraction of 0 or 1, and then pull the corresponding geometry to do a regular PIP test only if the fraction is in between.

**Comment by dr-jts** [comment]
If this is added to GEOS, I suggest a more descriptive name.  Something like `GridOverlay` or `GridIntersection`.

**Comment by theroggy** [comment]
I wonder if subdividing polygons up to a maximum number of points is a related algorithm that could benefit from the ideas here?
xref: #717

**Comment by dbaston** [comment]
@theroggy The "subdivide until no polygon has more than N points" formulation really lends itself to the recursive solution in PostGIS. This algorithm is useful in many of the same situations but the result is quite different.

It might be interesting to combine the two algorithms. Initial subdivision according to a grid and then recursive subdivision of the resulting non-rectangles.

**Comment by pramsey** [comment]
Isn't the main problem with grid subdivision the addition of new vertices into the data at the grid lines? @dr-jts keeps looking out of the side of his eyes at an approach that first triangulates, then builds up from there to the vertex limit by combining triangles. No new vertices, though potentially (a) a lot more CPU intensive? maybe? and (b) perhaps "ugly" resultants.

**Comment by dr-jts** [comment]
>  @dr-jts keeps looking out of the side of his eyes at an approach that first triangulates, then builds up from there to the vertex limit by combining triangles. No new vertices, though potentially (a) a lot more CPU intensive? maybe? and (b) perhaps "ugly" resultants.

Right, perhaps I should dust off this idea, now that RelateNG is done.

There's also the idea of adding "Steiner points" inside the polygon to make the triangulation more regular, and thus reduce the likelihood of having long pointy bits on the resultants.  Cf Shewchuk's [Triangle](https://www.cs.cmu.edu/~quake/triangle.html) program.

**Comment by theroggy** [comment]
> Isn't the main problem with grid subdivision the addition of new vertices into the data at the grid lines?

Probably just from my viewpoint for my needs, but I personally don't mind the extra points being added too much. there are definitely annoying side effects due to them... but you need to deal with those anyway when doing any overlays...

**Comment by pramsey** [comment]
Hm. I vaguely remember some complaints that a subdivide->union workflow might introduce slivers, but I cannot confirm or deny.

**Comment by strk** [comment]
> you need to deal with those anyway when doing any overlays...

Every single extra point can move segments introducing topological
errors, that's why NOT introducing new vertices is good for some
use cases.

**Comment by theroggy** [comment]
> you need to deal with those anyway when doing any overlays...
> Every single extra point can move segments introducing topological errors, that's why NOT introducing new vertices is good for some use cases.

Not at all to to say there aren't use cases where NOT introducing new vertices is an advantage... but I think it is interesting to further explore where topological errors could be introduced.

I don't have experience using the recursive subdivide, but at first sight I think the first step (just subdivide with a regular grid doesn't it introduces topological errors in a topologically correct layer yet? Because, when a polygon is split due to the layer being subdivided by a grid the extra points introduced will be added in both "new" split polygons, so they should stay topologically sound?

The recursions where certain tiles are further subdivided I think topological errors can be introduced if a neighbouring tile is not subdivided and the corner points of the further subdivided tiles are not added to the neighbouring polygons?

**Comment by dbaston** [comment]
> At first sight I think the first step (just subdivide with a regular grid doesn't it introduces topological errors in a topologically correct layer yet? Because, when a polygon is split due to the layer being subdivided by a grid the extra points introduced will be added in both "new" split polygons, so they should stay topologically sound?

This is what I'd expect as well, but I haven't thoroughly tried to make it fail. If you'd like to test a candidate implementation you can run the `subdivide` tool that builds with https://github.com/isciences/exactextract:

```
./subdivide  ~/data/tl_2022_us_county.shp --te -180 -90 180 90 --tr 1 1 /tmp/out.shp
```

**Comment by dr-jts** [comment]
> At first sight I think the first step (just subdivide with a regular grid doesn't it introduces topological errors in a topologically correct layer yet? Because, when a polygon is split due to the layer being subdivided by a grid the extra points introduced will be added in both "new" split polygons, so they should stay topologically sound?

Yes, theoretically this should be the case.  But computing line intersection has all kinds of tricky cases where things could wind up not matching, unless the algorithm is carefully designed to prevent this. For instance, I'm fairly sure that subdividing just using the OverlayNG algorithm might create resultants which don't match up exactly, due to various heuristics used which might affect one side but not the other.  Of course, more appropriate algorithms (such as `exactextract`) might not have this issue.  

> The recursions where certain tiles are further subdivided I think topological errors can be introduced if a neighbouring tile is not subdivided and the corner points of the further subdivided tiles are not added to the neighbouring polygons?

Yes, this could also be an issue.  One solution is to add all node vertices to adjacent polygons.  Of course, this is a more complicated algorithm.  Or potentially an algorithm can guarantee that nodes lie exactly on horizontal/vertical lines, which _should_ prevent slivers from occurring.  But it's always more reassuring to have edge-matched vertices.

**Comment by theroggy** [comment]
FWIW: I'm using the following implementation now to have a "poor man's subdivide", which is basically doing an intersection with a rectangular grid: 
- based on max coordinates wanted, create a grid in python with as bbox the bbox of the input polygon and tile size so if the points are perfectly distributed, the number of points would become < max coordinates.
- then intersect the grid with the input polygon using shapely (= GEOS).

Code:
https://github.com/pygeoops/pygeoops/blob/8ec1561f672656386d13239fed8088ad60d82ba3/pygeoops/_general.py#L476

**Comment by dbaston** [comment]
Implemented in #1295 and #1232
